File: Java/AddTwoNumbers.java
package main.java;

/**
 * 2. Add Two Numbers
 * You are given two linked lists representing two non-negative numbers. 
 * The digits are stored in reverse order and each of their nodes contain a single digit. 
 * Add the two numbers and return it as a linked list.

 * Input: (2 -> 4 -> 3) + (5 -> 6 -> 4)
 * Output: 7 -> 0 -> 8
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */

public class AddTwoNumbers {

	/**
	 * @param argss
	 */
	public ListNode addTwoNumbers(ListNode l1, ListNode l2){
		int carry = 0;
		int sum = 0;
		ListNode listRet = new ListNode(0);
		ListNode listTemp = listRet;
		
		if (null == l1) listTemp = l2;
		if (null == l2) listTemp = l1;
		
		while ((null != l1) && (null != l2)){	//两个链表一样长时
			sum = l1.val + l2.val + carry;
			carry = sum / 10;
			listTemp.next = new ListNode(sum % 10);
			listTemp = listTemp.next;
			l1 = l1.next;
			l2 = l2.next;
			
		}
		while (null != l1){		//两个链表不一样长时
			sum = l1.val + carry;
			carry = sum / 10;
			listTemp.next = new ListNode(sum % 10);
			listTemp = listTemp.next;
			l1 = l1.next;
		}
		while (null != l2){
			sum = l2.val + carry;
			carry = sum / 10;
			listTemp.next = new ListNode(sum % 10);
			listTemp = listTemp.next;
			l2 = l2.next;
		}
		if (carry == 1)		//最后一位有进位时
			listTemp.next = new ListNode(1);
		return listRet.next;	}
	
	public static void main(String[] args) {
		// TODO Auto-generated method stub

	}

}

File: C/addTwoNumbers.c
/*
2. 两数相加

给出两个 非空 的链表用来表示两个非负的整数。其中，它们各自的位数是按照 逆序 的方式存储的，并且它们的每个节点只能存储 一位 数字。

如果，我们将这两个数相加起来，则会返回一个新的链表来表示它们的和。

您可以假设除了数字 0 之外，这两个数都不会以 0 开头。

示例：

输入：(2 -> 4 -> 3) + (5 -> 6 -> 4)
输出：7 -> 0 -> 8
原因：342 + 465 = 807

*/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */

//刚开始写的超时，明显不合理，需要优化
struct ListNode* addTwoNumbers(struct ListNode* l1, struct ListNode* l2){
    struct ListNode *L1, *L2, *tmp;
    struct ListNode *res = NULL;

    L1 = l1;
    L2 = l2;

    int sum = 0;
    int index = 0;
    int multi, temp;
    
    while (L1 && L2) {
        multi = pow(10, index);
        if (L1) {
            sum += (L1->val * multi);
            L1 = L1->next;
        }
        if (L2) {
            sum += (L2->val * multi);
            L2 = L2->next;
        }
        index++;
    }

    if (sum == 0) {
        res = (struct ListNode*)malloc(sizeof(struct ListNode));
        res->val = 0;
        res->next = NULL;
        return res;
    }

    index--;

    while (sum) {
        multi = pow(10, index);
        //temp = sum % multi;
        temp = sum / multi;
        sum = sum % multi;

        tmp = (struct ListNode*)(void *)malloc(sizeof(struct ListNode));
        tmp->val = temp;
        tmp->next = NULL;

        if (!res) {
            res = tmp;
        } else {
            tmp->next = res;
            res = tmp;
        }
    }

    return res;
}



